Add CLI Script for Regeneration

RidRack opened the pull request:
Adds a shell script that calls the same function as the controller action to the extension, so that you can regenerate config from the command line.

Commit:
Add the regenerate php shell script and update the css model to allow for block creation

Adds a shell script that can be used to regenerate the config via the command line and fixes a problem with the css model that prevented using it from the command line
https://magento.stackexchange.com/a/58784

Commit:
Add the shell folder to the modman file.

tgerulaitis commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,48 @@
+            fwrite(STDERR,
+                "Config powered css has been regenerated successfully.\n");
+            exit(0);
+        } catch (Exception $e) {
```
What exceptions are we trying to catch here? This seems unnecessary.

RidRack replied:
I'd rather catch an exception then break a build process, as this is only generating css.

tgerulaitis replied:
Two problems with that:

1. Silently ignoring exceptions is bad practice. The deployment will go green and we'll happily move along only to discover later that the site had problems. If there are non-breaking exceptions, we can catch and alert, but catching **all** exceptions and ignoring them is bad code.
2. You have nested try-catch blocks with different behaviours. The actual "meat" of the work is wrapped in a try-catch which exits with an error code, so if something breaks with the CSS generation, we are actually going to stop the deployment. This try-catch block really only covers lines 21-24, where either we don't have the model we expect (bad, build breaking) or can't fetch stores (bad, build breaking).

My suggestion is to leave the deployment process to decide whether to stop the deployment if this script returns an error code, or carry on. The script itself should return an accurate report on the state of its changes.

tgerulaitis commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,48 @@
+                }
+            }
+
+            fwrite(STDERR,
```
We should not be writing success messages on standard error stream

tgerulaitis commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,48 @@
+                    $css->publish($storeId);
+                } catch (Exception $e) {
+                    Mage::logException($e);
+                    return;
```
This will stop on the first exception, but exit with no message to the user and a successful error code. We should show an error and use an error code to exit.

tgerulaitis commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,48 @@
+            Mage::logException($e);
+
+            fwrite(STDERR,
+                "There was an error when regenerating the config powered css, it has been logged.\n");
```
It would be better to show the error to the user, instead of making them jump through hoops for it (logging into the server, navigating to the log directory, finding the specific error in the endless stream of `exception.log`).

Commit:
Update the error messaging to be more useful and use the correct exit codes

RidRack commented:
I've updated this with the right changes, let me know what you think @tgerulaitis

bobbyshaw commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,47 @@
+                }
+            }
+
+            fwrite(STDOUT,
```
Line break formatting is weird here. As I understand it, if you break arguments onto multiple lines then every argument should be on its own line. Final closing brackets will be on their own line too, i.e.
```php
fwrite(
    STDOUT,
    "Your message\n"
);
```

bobbyshaw commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,47 @@
+                "Config powered css has been regenerated successfully.\n");
+            exit(0);
+        } catch (Exception $e) {
+            fwrite(STDERR,
```
New line formatting.

bobbyshaw commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,47 @@
+                    Mage::app()->setCurrentStore($storeId);
+                    $css->publish($storeId);
+                } catch (Exception $e) {
+                    fwrite(STDERR, sprintf('There was an error when regenerating the config powered css: %s',
```
New line formatting, see other comment.

bobbyshaw commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,47 @@
+            $css = Mage::getModel('meanbee_configpoweredcss/css');
+
+            $stores = Mage::app()->getStores();
+            foreach ($stores as $storeId => $store) {
```
I don't require it to be changed for this to be merged but this is heavily repeating code that's in the controller. That would tell me that it's worth pulling this for loop out into a function on a helper. The controller and this shell command would then just catch exceptions and handle them in their own way. Something to consider.

bobbyshaw commented:
It'd be nice to add a note to the readme on the new command use.

bobbyshaw reviewed (approved):
It'd be nice to add a note to the readme on the new command use.

Commit:
Refactor the Shell class and the Controller to use a helper,

allowing for them each to do their own error handling with less repetition of code

Commit:
Add the shell script usage to the readme file.

RidRack commented:
I think I've fixed all of the problems you both mentioned, @tgerulaitis Would you mind approving?

tgerulaitis commented on src/app/code/community/Meanbee/ConfigPoweredCss/controllers/Adminhtml/ConfigPoweredCssController.php:
```diff
@@ -7,20 +7,13 @@ public function regenerateAction()
+        try {
+            Mage::helper('meanbee_configpoweredcss')->regenerateConfigPoweredCss();
+        } catch (Exception $e) {
+            Mage::app()->setCurrentStore(Mage_Core_Model_App::ADMIN_STORE_ID);
```
I think this line belongs in the helper method, because of the division of responsibility. The helper method sets a different store during generations, so it should have the responsibility of resetting it afterwards. The controller shouldn't need the knowledge that the store could have/has been changed and needs to be reset.

Moreover, since the helper method can be used in any context, it should record the current store at the start and reset to it, instead of always resetting to the admin store. That's just protection against possible bugs in the future.

RidRack replied:
Fair point, I partially agree. I'll remove the later call as we shouldn't have to set the current store if  we've not had an exception, that can be done by the helper, however, in the case that we end up with an exception, it would be good that the controller sets the store to admin.

tgerulaitis replied:
I don't see how that improves things though. The controller shouldn't be responsible for resetting the environment in any case. The helper method shouldn't screw up the environment for the user, and leave side effects, whatever the user is. It's perfectly fine to catch the exception in the helper method, reset the environment, and then throw the exception forward again with `throw $e;`. In fact, the clearest change would be to use the `finally` keyword in the helper method, without catching the exception, i.e.:

```
try {
    // Do the work that messes with the environment
} finally {
    // Reset the environment
}
```

tgerulaitis commented on src/app/code/community/Meanbee/ConfigPoweredCss/controllers/Adminhtml/ConfigPoweredCssController.php:
```diff
@@ -7,20 +7,13 @@ public function regenerateAction()
+            return;
         }
 
         Mage::app()->setCurrentStore(Mage_Core_Model_App::ADMIN_STORE_ID);
```
Don't forget this line too, regarding the above comment

tgerulaitis commented on src/shell/meanbee/configpoweredcss/regenerate.php:
```diff
@@ -0,0 +1,31 @@
+            Mage::helper('meanbee_configpoweredcss')->regenerateConfigPoweredCss();
+        } catch (Exception $e) {
+            fwrite(STDERR, sprintf('There was an error when regenerating the config powered css: %s',
+                $e->getTraceAsString()));
```
Like Tom mentioned before, the convention is either to keep the entire method on one line, or split it to one argument per line, to ease readability.

RidRack replied:
Damn, my linter changed it after I'd fixed it and I didn't notice it.

Commit:
Set original store in helper, rather then in controller

Commit:
Add finally block to helper to always set the store even on exception
